libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -97,10 +97,39 @@ bool ContractLevelChecker::check(ContractDefinition const& _contract)
 	checkBaseABICompatibility(_contract);
 	checkPayableFallbackWithoutReceive(_contract);
 	checkStorageSize(_contract);
+	checkStorageLayoutSpecifier(_contract);
 
 	return !Error::containsErrors(m_errorReporter.errors());
 }
 
+void ContractLevelChecker::checkStorageLayoutSpecifier(ContractDefinition const& _contract)
+{
+	if (
+		(_contract.isLibrary() || _contract.abstract() || _contract.isInterface()) &&
```

Comment:
Since #15463 rejects libraries and interfaces during parsing now, you can just assert against them here. Abstract contracts are the only case that can reach type checking stage.